### ISSUE TITLE
Temporal Fusion Transformer gives constant prediction

### ISSUE THREAD

**Issue by manitadayon** [jdb78/pytorch-forecasting#1332]
- PyTorch-Forecasting version: 1.0.0
- PyTorch version: 2.0.1
- Python version: 3.10
- Operating System: Windows

### Expected behavior
My expectation is to get a prediction that is not completely constant and can capture some volatility in the curve.

### Actual behavior

No matter how much I tune hyperparameters, I get pure constant/linear prediction.
### Code to reproduce the problem

```
max_encoder_length = 100
max_prediction_length = 100

context_length = max_encoder_length
prediction_length = max_prediction_length

training = TimeSeriesDataSet(
    train_data,
    time_idx="new_time",
    target="units",
    group_ids=["marketplace"],
    time_varying_unknown_reals=["units"],
    max_encoder_length=context_length,
    max_prediction_length=prediction_length,
    allow_missing_timesteps=True,
    add_relative_time_idx=True,
    static_categoricals = ["marketplace"],
    target_normalizer = GroupNormalizer(groups = ['marketplace'], transformation = 'softplus')
)

batch_size = 128
train_dataloader = training.to_dataloader(train=True, batch_size=batch_size, num_workers=0)


net  = TemporalFusionTransformer.from_dataset(
training, learning_rate=0.1, hidden_size=27, attention_head_size=2, 
dropout=0.18, hidden_continuous_size=16, log_interval=1, log_val_interval = 1,output_size=7,
            loss=QuantileLoss(), lstm_layers = 3
)

trainer.fit(
    net,
    train_dataloaders=train_dataloader,
    val_dataloaders=val_dataloader
)

```
The result after prediction is:

```

for idx in range(5):
    best_model.plot_prediction(raw_predictions.x,raw_predictions.output, idx=idx)

```
![predicted_actual_TFT](https://github.com/jdb78/pytorch-forecasting/assets/19966010/cd1b454d-9632-4261-9882-26c79a7fef19)

Can someone help me understand what is wrong with my code? I have also tried it on different datasets and still had the same issue.

**Comment by meetri** [comment]
Same issue here. I did notice after including multiple categories in my data I got more variation in the prediction

**Comment by wkkg** [comment]
I have tried various combinations of hyperparameters and implementation from https://towardsdatascience.com/temporal-fusion-transformer-time-series-forecasting-with-deep-learning-complete-tutorial-d32c1e51cd91 the forecasts still remain flat. If I find the solution, I will let you know. Could you do the same? What about posting this problem on stackoverflow?

**Comment by meetri** [comment]
> Can you please elaborate on what you mean by including more categories? Are you talking about static categories? You mean more predictors? I can try it and update it here. 

Yes. Static categories. Curious what you end up finding. Looking forward to your update

**Comment by meteoDaniel** [comment]
We have the same problem. We started with pytorch-forecasting on a previous version, and kept going on with a custom implementation based on [playtika](https://www.playtika-blog.com/playtika-ai/multi-horizon-forecasting-using-temporal-fusion-transformers-a-comprehensive-overview-part-2/) . Training runs fine, and in-sample evaluation looks good. We have a very similar setup with pytorch and lightning > 2.0 . 
Most probably it is an issue with pytorch and/or lightning in v2.

**Comment by ManieTadayon** [comment]
@meteoDaniel, thanks this is very interesting, do you recommend downgrading version of pytorch lightening (seems to be suboptimal solution) have you tried darts?

**Comment by meteoDaniel** [comment]
@ManieTadayon we are actually trying this. For debugging it is a good idea to find the problem. We will keep you updated as soon as we have some helpful insights.

**Comment by JakeF-Bitweave** [comment]
Sorry I'm not close to the project anymore! Good luck

**Comment by grosestq** [comment]
> My expectation is to get a prediction that is not completely constant and can capture some volatility in the curve.

Hi,

is it really completely constant, as you said? Looking at your picture, I would say that the prediction is not constant.

Best regards

**Comment by ManieTadayon** [comment]
It learns absolutely no volatility compared to similar models regardless of features added or hyperparameter tuning. the change in forecast is minimal (hence almost constant).

**Comment by grosestq** [comment]
> It learns absolutely no volatility compared to similar models regardless of features added or hyperparameter tuning. the change in forecast is minimal (hence almost constant).

How does the prediction accuracy compare to similar models?

**Comment by valentinfrlch** [comment]
I am having the same problem here:

**Comment by valentinfrlch** [comment]
I am having the same problem here:
![prediction_B2F8KN-R_1](https://github.com/jdb78/pytorch-forecasting/assets/85313672/4238d118-f231-48cb-b518-4278c94307eb)
It seems to predict the average alright but no volatility at all. I am experimenting with encoder lenght but no dice so far.

Will let everyone know if I find a solution...

**Comment by valentinfrlch** [comment]
May I ask what other features you fed? Something like day of week, month etc. or more actual data?

**Comment by HHarald99** [comment]
Has there been any more insight in the meantime? Because I don't think adding features is the solution.

It is true that those add volatility, but this would still entirely come from those features like in a regression, not at all from the encoded input sequence. For example, when you have a weekly pattern and shift the input data by three days, all predictions would be off by those three days because the model did not pick up anything from encoder side and just repeated the statically learned weekday pattern. At least that's my experience even from the simplest toy datasets.

Maybe I have wrong expectations but I would have thought this kind of model can dynamically pick up patterns from the encoded sequence, like an ETS or Arima Model does. So I still wonder if this is a problem just in this implementation.

**Comment by valentinfrlch** [comment]
For me it was the normalizer. Softplus should be pretty good for most scenarios, I couldn't use it however since my numbers where too small at times. Check out [GroupNormalizer in the docs](https://pytorch-forecasting.readthedocs.io/en/stable/api/pytorch_forecasting.data.encoders.GroupNormalizer.html).
I had luck with relu but your mileage may vary.

**Comment by abudis** [comment]
@valentinfrlch for clarity, as you saying that _disabling_ GroupNormalizer is what did the trick for you?

**Comment by manitadayon** [comment]
@valentinfrlch, no by default TFT, NHITS, DeepAR, etc are normalizing the data sequence by sequence, you can configure it such that it is done by time series by time series.

**Comment by manitadayon** [comment]
@meteoDaniel Any update on your findings? I do not think it is the package.

**Comment by rd1886** [comment]
Hi,

Has anyone found the root cause of this and how to fix it? I got the same issue with the flat prediction. I have been trying to fix it for a while now.

**Comment by manitadayon** [comment]
To be honest I got some good prediction out of TFT. Try adding static variables, temporal variables, etc. However it still looks very suboptimal. For example, if you use NHiTS, you will get a good prediction without specifying much external covariates.

**Comment by rd1886** [comment]
Thank you, @manitadayon

**Comment by erwinvink** [comment]
Hi guys, I switched to Darts since its clear that pytorch-forecasting and TFT is not a good combination. I spend 500 hours of hyperparameter tuning, tried the most strange combinations but whatever I tried, it stays flat. I found this magnificant article but you have to build it yourself here: https://www.playtika-blog.com/playtika-ai/multi-horizon-forecasting-using-temporal-fusion-transformers-a-comprehensive-overview-part-2/
So I switched to Darts and it was working instantly with the most simple training.: https://unit8co.github.io/darts/examples/13-TFT-examples.html

**Comment by manitadayon** [comment]
@erwinvink thank you. However I am confused, did you end up using darts or tft-PyTorch: since the article shows tft-pytorch packages and then you mentioned using darts again. 
did you use steps outlined there only for data preprocessing and then darts for training?
May I know why not the package itself for the whole thing.

**Comment by erwinvink** [comment]
@manitadayon : I ended up using darts. Apologies for the confusion. I see that my message is indeed not clear ;). I mentioned the article so you can choose yourself if you want to use tft-pytorch or darts. But I use darts for myself. The data preprocessing part (getting data, transforming, reverse transforming) was working already so I didn't need to change that part. I use darts to train the model itself and make predictions.

**Comment by Rujie-Zhu** [comment]
I also faced the same problem. My input features have different units and magnitudes. I tried to keep the features with the same units of the target and remove the others. The prediction can capture the volatility. So I guess it is something wrong with the preprocess of the data.

**Comment by t4ai** [comment]
Also having this issue - I have tried tuning, changing the GroupNormalizer adding parameters, per all the great advice above and still no luck.  Would love to hear if anyone has found a solution. It sounds like Darts works better but I am hesitant to make the jump.

**Comment by sanderlod** [comment]
We are also experiencing this issue, and Dart did not solve the problem for us.  I'd be interested to know if anyone has identified the root cause or discovered any effective solutions.

**Comment by erwinvink** [comment]
@sanderlod : If Darts gives you the same problem, I would say it's because of your scalers. What data do you try to predict and which scalers do you use when preprocessing the data?

**Comment by manitadayon** [comment]
For folks that use darts successfully, how do you feed multiple time series with overlapping time to darts. In Pytorch forecasting this is super easy.

**Comment by erwinvink** [comment]
@manitadayon : This is a part of the setup of my model:

```py
tft_model = TFTModel(
                model_name=model_name,
                work_dir=workdir,
                force_reset=True,
                save_checkpoints=True,
                show_warnings=True,
                input_chunk_length=sweep_config['max_encoder_length'],
                output_chunk_length=sweep_config['max_decoder_length'],
                hidden_size=sweep_config['hidden_size'],
                hidden_continuous_size=sweep_config['hidden_continuous_size'],
                lstm_layers=sweep_config['lstm_layers'],
                num_attention_heads=sweep_config['num_attention_heads'],
                dropout=sweep_config['dropout'],
                batch_size=sweep_config['batch_size'],
                n_epochs=self.model_config.MODEL_CONFIG['max_epochs'],
                add_relative_index=False,
                add_encoders={
                    'cyclic': {'past': ['hour', 'dayofweek', 'week', 'month'], 'future': ['hour', 'dayofweek', 'week', 'month']},
                    'datetime_attribute': {'past': ['minute', 'hour', 'day'], 'future': ['minute', 'hour', 'day', 'week', 'month', 'year']},
                    'position': {'past': ['relative'], 'future': ['relative']},
                    'transformer': Scaler(),
                    'tz': 'CET'
                },
                lr_scheduler_cls=ReduceLROnPlateau,
                lr_scheduler_kwargs={
                    "optimizer": torch.optim.Adam,
                    "mode": 'min',
                    "factor": 0.5,
                    "patience": sweep_config['reduce_on_plateau_patience'],
                    "verbose": True
                },
```

As you can see, I have added the timeseries here in an encoder (according to the somewhat limited documentation in darts itself: https://unit8co.github.io/darts/generated_api/darts.dataprocessing.encoders.encoders.html?highlight=cyclic+encoder#sequentialencoder

After this, it was much easier for me to added series with overlapping timedata. 

```
model = tft_model.fit(
                series=self.data['train_target'],
                past_covariates=self.data['train_past_covariates'],
                val_series=self.data['val_target'],
                val_past_covariates=self.data['val_past_covariates'],
                verbose=True,
                num_loader_workers=self.model_config.MODEL_CONFIG["num_workers"]
            )
```
Is this helping you?

**Comment by manitadayon** [comment]
@erwinvink Thanks a lot, this is super useful. What I really need to know is: how does `self.data['train_target']` look like?
Is it a panda dataframe for multiple time series/products/ID concatenate across rows where each time series might have overlapping time frames. For example lets say we have 30 rows for product A from 2020-03-01 to 2020-04-01 and product B is from 2020-02-01 to 2021-02-01.

**Comment by erwinvink** [comment]
@ManieTadayon : hereby the function where I build my data (which is a list with multiple timeseries in it):

```py
    def create_timeseries(self, df):
        # Create the target series
        target = TimeSeries.from_times_and_values(times=df.index, values=df['close'])
        train_target, val_target = target.split_before(0.8)

        # Create the unknown covariates
        past_covariates = TimeSeries.from_times_and_values(times=df.index,
                                                           values=df[self.config.MODEL_CONFIG['features']],
                                                           columns=self.config.MODEL_CONFIG['features'])
        train_past_covariates, val_past_covariates = past_covariates.split_before(0.8)

        data = {}
        data['target'] = target
        data['train_target'] = train_target
        data['val_target'] = val_target
        data['past_covariates'] = past_covariates
        data['train_past_covariates'] = train_past_covariates
        data['val_past_covariates'] = val_past_covariates

        return data
```

the `self.config.MODEL_CONFIG['features']` is just a list with the column names I want to include here.

**Comment by manitadayon** [comment]
@erwinvink Thanks a lot for this. You mention you data is a list of multiple time series. Here df is the dataframe and data that you return is dictionary.  Is what you feed to the `create_timeseries` function a  list of dataframe and you apply the `create_timeseries` function in a loop to each element of list and return data for each time series back and you concatenate (each data to be dataframe again, like may I know what is the final data you feed to `create_timeseries` and the one you feed to `TFTModel`)

**Comment by WaqasNaqvi03** [comment]
Guys if you mean the model is predicting the same constant values for the next steps then try setting the allow_missing_timesteps=False

Setting it to true forces the model to repeat the values for the missing time step which of course impacts the predictions.
